server/src/main/java/org/elasticsearch/common/time/JavaDateFormatter.java

```diff
@@ -22,10 +22,43 @@
 import java.util.List;
 import java.util.Locale;
 import java.util.Objects;
-import java.util.function.Consumer;
+import java.util.function.BiConsumer;
 import java.util.function.UnaryOperator;
 
 class JavaDateFormatter implements DateFormatter {
+    /**
+    * A default consumer that allows to round up fields (used for range searches, optional fields missing)
+    * it relies on toString implementation of DateTimeFormatter and ChronoField.
+    * For instance for pattern
+    * the parser would have a toString()
+    * <code>
+    *  Value(MonthOfYear,2)'/'Value(DayOfMonth,2)'/'Value(YearOfEra,4,19,EXCEEDS_PAD)'
+    * 'Value(ClockHourOfAmPm,2)':'Value(MinuteOfHour,2)' 'Text(AmPmOfDay,SHORT)
+    * </code>
+    * and ChronoField.CLOCK_HOUR_OF_AMPM would have toString() ClockHourOfAmPm
+    * this allows the rounding logic to default CLOCK_HOUR_OF_AMPM field instead of HOUR_OF_DAY
+    * without this logic, the rounding would result in a conflict as HOUR_OF_DAY would be missing, but CLOCK_HOUR_OF_AMPM would be provided
+    */
+    private static final BiConsumer<DateTimeFormatterBuilder, DateTimeFormatter> DEFAULT_ROUND_UP = (builder, parser) -> {
+        if (parser.toString().contains(ChronoField.DAY_OF_YEAR.toString())) {
```

Comment:
Just a small comment here, since we might call this frequently enough and we don't know the cost of parser.toString(), do you mind extracting the parsed value into a local for the if conditionals?

Reply:
good idea